### PR TITLE
feat(Core): gracefully shutdown router/scheduler if Redis is unavailable

### PR DESCRIPTION
**Pull request by prajjwalkumar17**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates

## Description
The change will from now allow the server to gracefully shutdown once the redis goes down. We are passing the tx and rx using a Oneshot::channel() for redis . It has a suitable receiver to stop the server once the tx is received.

### Additional Changes

- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Motivation and Context
As the core Server wasn't shutting down whenever redis goes down and rather that going down with it. Server stayed there throwing health is good and then sending errors when hitting other routes.
Now the server gracefully goes down too once the redis goes down

## How did you test it?
Tested it manually as we just need to check whether or not the server is shutting down on redis stopping

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [ ] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

**Comment by NishantJoshi00** [comment]
Hey @prajjwalkumar17, please setup commit verification and enforce it through github

File: crates/common_utils/src/pii.rs [hunk 1]
```diff
@@ -152,7 +152,8 @@ mod pii_masking_strategy_tests {
     #[test]
     fn test_invalid_card_number_masking() {
         let secret: Secret<String, CardNumber> = Secret::new("1234567890".to_string());
-        assert_eq!("123456****", format!("{secret:?}"));
+        println!("//////////////////////////////////////{:?}", secret);
```
Comment (NishantJoshi00) [inline crates/common_utils/src/pii.rs:155]: remove this, if not required

File: crates/drainer/src/lib.rs [hunk 2]
```diff
@@ -20,6 +21,7 @@ pub async fn start_drainer(
     max_read_count: u64,
     shutdown_interval: u32,
     loop_interval: u32,
+    // (tx,rx):(mpsc::Sender<()>,&mut mpsc::Receiver<()>),
```
Comment (NishantJoshi00) [inline crates/drainer/src/lib.rs:24]: remove comments

File: crates/redis_interface/src/lib.rs [hunk 3]
```diff
@@ -138,10 +138,16 @@ impl RedisConnectionPool {
             };
         }
     }
-    pub async fn on_error(&self) {
-        while let Ok(redis_error) = self.pool.on_error().recv().await {
+
+    pub async fn on_error(&self, option_tx: Option<tokio::sync::oneshot::Sender<()>>) {
```
Comment (NishantJoshi00) [inline crates/redis_interface/src/lib.rs:142]: - Does this tx need to be optional?
- Please don't prefix variable name with type information, we aren't following this conversion.

File: crates/router/src/scheduler/consumer.rs [hunk 4]
```diff
@@ -82,14 +84,17 @@ pub async fn start_consumer(
                     workflow_selector,
                 ));
             }
-            Ok(()) | Err(oneshot::error::TryRecvError::Closed) => {
+            Ok(()) | Err(mpsc::error::TryRecvError::Disconnected) => {
                 logger::debug!("Awaiting shutdown!");
+                ////can panic!
+                rx.close();
```
Comment (NishantJoshi00) [inline crates/router/src/scheduler/consumer.rs:90]: Not sure, if this panics. Can you review the docs and remove this comment, if not necessary

File: crates/router/src/scheduler/consumer.rs [hunk 5]
```diff
@@ -82,14 +84,17 @@ pub async fn start_consumer(
...
                 shutdown_interval.tick().await;
                 let active_tasks = consumer_operation_counter.load(atomic::Ordering::Acquire);
 
                 match active_tasks {
                     0 => {
                         logger::info!("Terminating consumer");
+                        rx.close();
```
Comment (NishantJoshi00) [inline crates/router/src/scheduler/consumer.rs:97]: This won't be required, as the channel has been closed already

**Review by NishantJoshi00 — CHANGES_REQUESTED** [review]

File: crates/router/src/scheduler/consumer.rs (hunk 5 above)
Comment (prajjwalkumar17) [reply crates/router/src/scheduler/consumer.rs:97]: On it

File: crates/redis_interface/src/lib.rs (hunk 3 above)
Comment (prajjwalkumar17) [reply crates/redis_interface/src/lib.rs:142]: I removed the Options as the method is called only at one place moreover I refactored the variable name

File: crates/router/src/scheduler/consumer.rs (hunk 4 above)
Comment (prajjwalkumar17) [reply crates/router/src/scheduler/consumer.rs:90]: Not required as it is closed already

File: crates/router/src/lib.rs [hunk 6]
```diff
@@ -149,10 +151,41 @@ pub async fn start_server(conf: settings::Settings) -> ApplicationResult<(Server
...
+#[async_trait::async_trait]
+impl Stop for mpsc::Sender<()> {
+    async fn stop_server(&mut self) {
+        match self.send(()).await {
+            Ok(_) => {}
+            Err(e) => logger::error!("{e}"),
+        }
+    }
+}
+
```
Comment (dracarys18) [inline crates/router/src/lib.rs:188]: This is not needed for now?

File: crates/router/src/bin/scheduler.rs [hunk 7]
```diff
@@ -18,14 +19,16 @@ async fn main() -> CustomResult<(), errors::ProcessTrackerError> {
     #[allow(clippy::expect_used)]
     let conf = Settings::with_config_path(cmd_line.config_path)
         .expect("Unable to construct application configuration");
-
-    let mut state = routes::AppState::new(conf).await;
+    let (oneshot_tx, oneshot_rx) = oneshot::channel();
+    let mut state = routes::AppState::new(conf, Some(oneshot_tx)).await;
+    let (tx, rx) = mpsc::channel(1);
+    tokio::spawn(router::receiver_for_error(oneshot_rx, tx.clone()));
```
Comment (dracarys18) [inline crates/router/src/bin/scheduler.rs:25]: Please add comments here

File: crates/router/Cargo.toml [hunk 8]
```diff
@@ -51,6 +51,7 @@ http = "0.2.9"
 josekit = "0.8.2"
 jsonwebtoken = "8.2.0"
 literally = "0.1.3"
+lazy_static = "1.4.0"
```
Comment (dracarys18) [inline crates/router/Cargo.toml:54]: Remove this

**Review by dracarys18 — CHANGES_REQUESTED** [review]

File: crates/common_utils/src/pii.rs (hunk 1 above)
Comment (prajjwalkumar17) [reply crates/common_utils/src/pii.rs:155]: Done manually added `*** alloc::string::String ***` as its for invalid card nos

File: crates/redis_interface/src/lib.rs [hunk 9]
```diff
@@ -138,10 +138,16 @@ impl RedisConnectionPool {
...
+        if let Ok(redis_error) = self.pool.on_error().recv().await {
```
Comment (SanchithHegde) [inline crates/redis_interface/src/lib.rs:143]: This would have to be `while let`.

File: crates/router/src/lib.rs [hunk 10]
```diff
@@ -149,10 +151,41 @@ pub async fn start_server(conf: settings::Settings) -> ApplicationResult<(Server
         .workers(server.workers)
         .shutdown_timeout(server.shutdown_timeout)
         .run();
-
+    tokio::spawn(receiver_for_error(rx, server.handle()));
     Ok((server, app_state))
 }
 
+pub async fn receiver_for_error(rx: oneshot::Receiver<()>, server: impl Stop) {
```
Comment (dracarys18) [inline crates/router/src/lib.rs:158]:
```suggestion
pub async fn receiver_for_error(rx: oneshot::Receiver<()>, mut server: impl Stop) {
```
and remove line number `162`.

**Review by dracarys18 — CHANGES_REQUESTED** [review]

File: crates/router/src/lib.rs [hunk 11]
```diff
@@ -149,10 +151,41 @@ pub async fn start_server(conf: settings::Settings) -> ApplicationResult<(Server
...
+    match rx.await {
+        Ok(_) | Err(_) => {
```
Comment (dracarys18) [inline crates/router/src/lib.rs:160]: Just log the error if receiver throws any. We need to stop the server only in the Ok(_) case

File: crates/router/src/lib.rs [hunk 12]
```diff
@@ -149,10 +151,41 @@ pub async fn start_server(conf: settings::Settings) -> ApplicationResult<(Server
...
+        let _ = self.stop(true).await;
+    }
+}
+#[async_trait::async_trait]
+impl Stop for mpsc::Sender<()> {
+    async fn stop_server(&mut self) {
+        match self.send(()).await {
+            Ok(_) => {}
+            Err(e) => logger::error!("{e}"),
+        }
```
Comment (dracarys18) [inline crates/router/src/lib.rs:185]:
```suggestion
       self.send(()).await.map_err(|err| logger::error!("{e}");
```

File: crates/drainer/src/lib.rs [hunk 13]
```diff
@@ -35,14 +36,15 @@ pub async fn start_drainer(
             .change_context(errors::DrainerError::SignalError(
                 "Failed while getting allowed signals".to_string(),
             ))?;
-    let (tx, mut rx) = oneshot::channel();
+
+    let (tx, mut rx) = tokio::sync::mpsc::channel(1);
```
Comment (dracarys18) [inline crates/drainer/src/lib.rs:40]:
```suggestion
    let (tx, mut rx) = mpsc::channel(1);
```

Comment (dracarys18) [inline crates/drainer/src/lib.rs:40]:
```suggestion
    let (tx, mut rx) = mpsc::channel(1);
```

**Review by SanchithHegde — APPROVED** [review]

File: crates/router/src/bin/scheduler.rs [hunk 14]
```diff
@@ -18,14 +19,20 @@ async fn main() -> CustomResult<(), errors::ProcessTrackerError> {
     #[allow(clippy::expect_used)]
     let conf = Settings::with_config_path(cmd_line.config_path)
         .expect("Unable to construct application configuration");
-
-    let mut state = routes::AppState::new(conf).await;
+    //we are having two channels here in order to close redis_interface & scheduler using oneshot and closing drainer using mpsc channel
+    let (redis_shutdown_signal_tx, redis_shutdown_signal_rx) = oneshot::channel();
+    let mut state = routes::AppState::new(conf, redis_shutdown_signal_tx).await;
+    let (tx, rx) = mpsc::channel(1);
```
Comment (NishantJoshi00) [inline crates/router/src/bin/scheduler.rs:25]:
```suggestion
    // channel for listening to redis disconnect events
    let (redis_shutdown_signal_tx, redis_shutdown_signal_rx) = oneshot::channel();
    let mut state = routes::AppState::new(conf, redis_shutdown_signal_tx).await;
    // channel to shutdown scheduler gracefully
    let (tx, rx) = mpsc::channel(1);
```

**Review by NishantJoshi00 — APPROVED** [review]

**Review by dracarys18 — APPROVED** [review]

**Review by SanchithHegde — APPROVED** [review]

**Review by NishantJoshi00 — DISMISSED** [review]

**Review by dracarys18 — DISMISSED** [review]

File: crates/redis_interface/Cargo.toml [hunk 15]
```diff
@@ -12,8 +12,9 @@ async-trait = "0.1.68"
 error-stack = "0.3.1"
 fred = { version = "6.0.0", features = ["metrics", "partial-tracing"] }
 futures = "0.3"
-serde = { version = "1.0.160", features = ["derive"] }
-thiserror = "1.0.40"
+serde = { version = "1.0.155", features = ["derive"] }
+thiserror = "1.0.39"
```
Comment (SanchithHegde) [inline crates/redis_interface/Cargo.toml:16]: Please use newer versions of these dependencies.

**Review by SanchithHegde — DISMISSED** [review]

**Review by jarnura — COMMENTED** [review]
Address the import changes, other changes are looks good

File: crates/router/src/services.rs [hunk 16]
```diff
@@ -94,22 +95,24 @@ pub(crate) struct StoreConfig {
 }
 
 impl Store {
-    pub async fn new(config: &crate::configs::settings::Settings, test_transaction: bool) -> Self {
+    pub async fn new(
+        config: &crate::configs::settings::Settings,
```
[the remark this answers is not in the document]
Comment (prajjwalkumar17) [reply crates/router/src/services.rs]: I guess this is creating some issues what is the suggestion here

**Review by dracarys18 — DISMISSED** [review]

**Review by NishantJoshi00 — DISMISSED** [review]

**Review by jarnura — APPROVED** [review]

**Review by SanchithHegde — APPROVED** [review]

**Review by jarnura — APPROVED** [review]

**Pull request merged**